src/auxinfo/proof.rs

```diff
@@ -143,3 +143,193 @@ impl AuxInfoProof {
+        let shared_context = &generate_shared_context();
+        let mut transcript = AuxInfoProof::new_transcript();
+        AuxInfoProof::append_pimod_transcript(&mut transcript, shared_context, sid, rho)?;
+        let pimod = pimod::PiModProof::prove(
+            &pimod::CommonInput::new(&modulus),
+            &pimod::ProverSecret::new(&p, &q),
+            shared_context,
+            &mut transcript,
+            &mut rng,
+        )?;
+        AuxInfoProof::append_pifac_transcript(&mut transcript, shared_context, sid, rho)?;
+        let pifac = pifac::PiFacProof::prove(
+            pifac::CommonInput::new(&setup_params, &modulus),
+            pifac::ProverSecret::new(&p, &q),
+            shared_context,
+            &mut transcript,
+            &mut rng,
+        )?;
+        let proof = AuxInfoProof {
+            pimod: pimod.clone(),
+            pifac: pifac.clone(),
+        };
+        assert!(proof
+            .verify(shared_context, sid, rho, &setup_params, &modulus)
+            .is_ok());
+        let mut rng1 = init_testing();
+        let sid1 = Identifier::random(&mut rng1);
+        let rho1: [u8; 32] = random();
+        let setup_params1 = VerifiedRingPedersen::gen(&mut rng1, &())?;
+        let (p1, q1) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng1).unwrap();
+        let modulus1 = &p1 * &q1;
+        let shared_context1 = &generate_shared_context();
+        let mut transcript1 = AuxInfoProof::new_transcript();
+        AuxInfoProof::append_pimod_transcript(&mut transcript1, shared_context1, sid1, rho1)?;
+        let pimod1 = pimod::PiModProof::prove(
+            &pimod::CommonInput::new(&modulus1),
+            &pimod::ProverSecret::new(&p1, &q1),
+            shared_context1,
+            &mut transcript1,
+            &mut rng1,
+        )?;
+        AuxInfoProof::append_pifac_transcript(&mut transcript1, shared_context1, sid1, rho1)?;
+        let pifac1 = pifac::PiFacProof::prove(
+            pifac::CommonInput::new(&setup_params1, &modulus1),
+            pifac::ProverSecret::new(&p1, &q1),
+            shared_context1,
+            &mut transcript1,
+            &mut rng1,
+        )?;
+        let proof1 = AuxInfoProof {
+            pimod: pimod1.clone(),
+            pifac: pifac1.clone(),
+        };
+        assert!(proof1
+            .verify(shared_context1, sid1, rho1, &setup_params1, &modulus1)
+            .is_ok());
+        let proof2 = AuxInfoProof {
+            pimod,
+            pifac: pifac1,
+        };
+        assert!(proof2
+            .verify(shared_context1, sid1, rho1, &setup_params1, &modulus1)
+            .is_err());
+        let proof3 = AuxInfoProof {
+            pimod: pimod1,
+            pifac,
+        };
+        assert!(proof3
+            .verify(shared_context1, sid1, rho1, &setup_params1, &modulus1)
+            .is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn pimod_pifac_proofs_must_be_correct_individually() -> Result<()> {
```

Comment:
Again, there's a lot of duplicated code here. The reason these proofs are failing is because the transcript is wrong, though. The test name implies that it's because the proofs are wrong, but both proofs are validly constructed.

1. I think testing that the transcript has to be correct is a useful thing to have, but I would rather do that via the actual `AuxInfoProof` APIs, e.g. by passing a different transcript to `prove` and `verify` and making sure it fails.

2. My original intention was to use the `AuxInfoProof::prove` method to create a valid proof and then manually changing a field e.g.
```rust
let mut proof = AuxInfoProof::prove(...);
proof.pifac.p_commitment = /* garbage */;
assert!(proof.verify(...).is_err());
```
However, I now realize that won't work because we made a good API for the other proofs and their fields cannot be publicly accessed and changed like this.

Instead, I think we can just pass bad inputs to the `AuxInfoProof::prove` method, like `p` and `q` that don't multiply to the modulus. As with pifac, this will probably be a test that might fail on proving and might fail on verifying.
    
A separate test could be to do different common inputs for proving and verifying, different shared contexts for proving and verifying, and, as I suggested in 1, different transcripts for proving and verifying.